Rework code calculating memory usage - output and Docker container

Malikbadmus opened the pull request:
**Description**

Rework code calculating memory usage - output and Docker container.

**Summary**

Ersilia model hub has different model server Implementations, each implementation handles specific deployment scenarios, such as local installation, conda environments, Docker containers, hosted services, etc. These implementations are defined in `Autoservice` class which makes use of the `service_class` objects to specify what service implementation to use for serving a model. By default, Ersilia models are fetched through Dockerhub and executed in a container. 

From this, we can split our model into two (2) groups.

- Models executed in a Docker Container.
- Models that do not utilize a Docker container.

The task here is to get three (3) key memory metrics from the underlying OS for these two (2) model types, for models that are executed in docker containers we will gather six (6) memory metrics.

- Total memory being used by the Docker container running the model.
- Peak memory used by the Docker container running the model
- CPU time used by the Docker container running the model.
- Total physical memory (ROM/RAM) being used at the serving and run level.
- Peak memory (ROM/RAM) being used at the serving and run level
- CPU time being used at the serving and run level.

For the second model type, we are only interested in the last three memory metrics listed above, to get these memory statistics, we made use of different libraries and APIs, specifically, Docker has a list of APIs for capturing such metrics, and apart from the key `max_usage` for getting the peak memory, which only works for a system with **cgroup V1** . To address this limitation, a solution has been included to access container memory data directly from the cgroup filesystem.

Also in the Ersilia process, after a model has been served through the model server, the memories are deallocated, so there is a need to update the memory information across different sessions(Serving, Getting predictions).   

 **USS** (unique set size) was used to get the total memory being used as opposed to **RSS** (resident set size), as it is the most representative metric for determining how much memory is being used by a process.

**Changes Made**

- Added  `docker_stats` method to get the different memory stats of  a docker, which will return a message if the process is not being executed in a container.
- Remove  `start_tracking` method as we are not utilizing tracemalloc.
- Modify `peak_memory`  method to get the peak memory usage of the process running the model.
- Added `get_memory_info` method that utilizes `psutil` to get the CPU time and the total memory used by the actual model.
- Added  `update_cpu_time` and `update_total_memory` functions in `session.py` to manage data accumulated across sessions. These functions continuously update CPU time and total memory values as they change during various sessions. Additionally, introduced `update_peak_memory` to track and compare peak memory values across sessions.
- Modify `serve` and `run` in `model.py` to reflect this changes and to ensure tracking and the various tracking directories are only being created when the track flags are used.

**Status**

- Various memory metrics tracked and gotten across a model Runtime. 

- The logging folder is only created and tracking is only done when a track flag is used.


Related to https://github.com/ersilia-os/ersilia/issues/1090 #1162

Malikbadmus commented:
@DhanshreeA , changes incoporated into log and session file here.

[eos3b5e.log](https://github.com/user-attachments/files/15866573/eos3b5e.log)
[2024-06-17_12-45-51.txt](https://github.com/user-attachments/files/15866580/2024-06-17_12-45-51.txt)

Malikbadmus commented:
> Hey @Malikbadmus good work! Could you resolve the conflicts in your PR so I can further review it? :) Thanks a lot!

Done!! Also formatted the code with Black.

Malikbadmus commented:
@DhanshreeA , the issue with the `log_file_metric` method has been fixed, now all the error's and warning's encountered throughout the model runtime will be logged once the Ersilia close command is used.

Also fixes this bug #1155

I have ran several test with and without the track flag and everything works. This is how the tracking file looks now.

[2024-06-18_11-43-25.txt](https://github.com/user-attachments/files/15884890/2024-06-18_11-43-25.txt)

DhanshreeA commented on ersilia/core/tracking.py:
```diff
@@ -146,28 +209,31 @@ def write_persistent_file(contents, model_id):
     else:
-        raise FileNotFoundError(f"The persistent file for model {model_id} does not exist. Cannot close file.")
-        
+        raise FileNotFoundError(
```
Do we need a blank line here? :)

DhanshreeA commented on ersilia/core/tracking.py:
```diff
@@ -3,26 +3,85 @@
-TEMP_FILE_LOGS = os.path.abspath("")
+def docker_stats(container_name=None):
+    """
+    This function will calculate the memory usage of the Docker container running Ersilia Models.
```
This docstring is misleading. You are calculating both memory and cpu consumption which is not reflected here. :)

DhanshreeA commented on ersilia/core/tracking.py:
```diff
@@ -3,26 +3,85 @@
 
-# Temporary path to log files until log files are fixed
-TEMP_FILE_LOGS = os.path.abspath("")
+def docker_stats(container_name=None):
```
I think the `tracking.py` module is growing phenomenally large so we need to move this code to a more suitable module. I recommend moving it to https://github.com/ersilia-os/ersilia/blob/master/ersilia/utils/docker.py where you create two new methods in the `SimpleDocker` class as follows: `container_memory`, and `container_cpu`, and then use those here. And then you can create an object of this class in RunTracker and use these methods. Since the class doesn't really rely on any initial conditions, this should be straightforward.

DhanshreeA commented:
> @DhanshreeA , changes incoporated into log and session file here.
> 
> [eos3b5e.log](https://github.com/user-attachments/files/15866651/eos3b5e.log) [current_session.txt](https://github.com/user-attachments/files/15866654/current_session.txt)
Hey @Malikbadmus  just confirming does the file `eos3b5e.log` contain memory and cpu stats from the ersilia CLI process? And the current_session.txt file contains those values for the container?

Malikbadmus commented:
> > @DhanshreeA , changes incoporated into log and session file here.
> > [eos3b5e.log](https://github.com/user-attachments/files/15866651/eos3b5e.log) [current_session.txt](https://github.com/user-attachments/files/15866654/current_session.txt)
> 
> Hey @Malikbadmus just confirming does the file `eos3b5e.log` contain memory and cpu stats from the ersilia CLI process? And the current_session.txt file contain those values for the container?

Yes,

Malikbadmus commented:
"Hi @DhanshreeA, I've reverted the last three changes and created a separate PR for them. Additionally, I've implemented all the suggested changes

DhanshreeA commented:
> > > @DhanshreeA , changes incoporated into log and session file here.
> > > [eos3b5e.log](https://github.com/user-attachments/files/15866651/eos3b5e.log) [current_session.txt](https://github.com/user-attachments/files/15866654/current_session.txt)
> > 
> > 
> > Hey @Malikbadmus just confirming does the file `eos3b5e.log` contain memory and cpu stats from the ersilia CLI process? And the current_session.txt file contain those values for the container?
> 
> Yes, I figured, since we are tracking memory stats between two sessions, we should have it in the log file. should i change it to the current session file?

How do you mean "between two sessions"? Do you mean between two processes?

Malikbadmus commented:
> > > > @DhanshreeA , changes incoporated into log and session file here.
> > > > [eos3b5e.log](https://github.com/user-attachments/files/15866651/eos3b5e.log) [current_session.txt](https://github.com/user-attachments/files/15866654/current_session.txt)
> > > 
> > > 
> > > Hey @Malikbadmus just confirming does the file `eos3b5e.log` contain memory and cpu stats from the ersilia CLI process? And the current_session.txt file contain those values for the container?
> > 
> > 
> > Yes, I figured, since we are tracking memory stats between two sessions, we should have it in the log file. should i change it to the current session file?
> 
> How do you mean "between two sessions"? Do you mean between two processes?

Yes, I meant processes(Model serving and inference).

DhanshreeA commented:
Hi @Malikbadmus my apologies again, but could you resolve the conflicts in this branch? It's good to go and I'll merge it right away when that's done. Thank you!

Malikbadmus commented:
Hello @DhanshreeA , the conflicts has been resolved.
